### PR TITLE
feat: download job manifests

### PR DESCRIPTION
**Pull request by andrew-fenton**
### What was the problem/requirement? (What/Why)
Part of the bucket sweeper project.

This PR implements functions to list and download all manifests pertaining to a job.

The manifest S3 key listing and download functions are separated to allow the bucket sweeper to access and delete the manifest files as well as the assets.

### What was the solution? (How)
1. List the S3 keys of all input and output manifests for a given job.
2. Merge the manifest keys into a single list.
3. Download all manifests by using their S3 keys.

**Example usage:**
```
 # Your IDs to query
farm_id = ""
queue_job_id_map: Dict[str, List[str]] = {}

for queue_id, job_ids in queue_job_id_map.items():
    for job_id in job_ids:
        manifest_keys: List[str] = _get_all_manifest_s3_keys_for_job(..., farm_id, queue_id, job_id)
        _download_job_manifests_using_s3_keys(..., manifest_keys)
```

### What is the impact of this change?
No impact to DeadlineCloud core functionality. These changes are read only, they do not modify or mutate any Deadline data.

### How was this change tested?
- Added unit tests
- Ran unit tests
- Ran integration tests
- Manual E2E test with service account

See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#testing) for information on running tests.

- Have you run the unit tests?
Screenshot 2025-07-10 at 4 41 52 PM

- Have you run the integration tests?
Screenshot 2025-07-10 at 4 40 52 PM

- Have you made changes to the `download` or `asset_sync` modules? If so, then it is highly recommended
  that you ensure that the docker-based unit tests pass. No

### Was this change documented?

- Are relevant docstrings in the code base updated?
Yes

- Has the README.md been updated? If you modified CLI arguments, for instance.
No changes necessary.

### Does this PR introduce new dependencies?

This library is designed to be integrated into third-party applications that have bespoke and customized deployment environments. Adding dependencies will increase the chance of library version conflicts and incompatabilities. Please evaluate the addition of new dependencies. See the [Dependencies](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies) section of DEVELOPMENT.md for more details.

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies).
*   [x] This PR does not add any new dependencies.

### Is this a breaking change? No

A breaking change is one that modifies a public contract in a way that is not backwards compatible. See the 
[Public Contracts](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#public-contracts) section
of the DEVELOPMENT.md for more information on the public contracts.

If so, then please describe the changes that users of this package must make to update their scripts, or Python applications.

### Does this change impact security? No

- Does the change need to be threat modeled? For example, does it create or modify files/directories that must only be readable by the process owner?
    - If so, then please label this pull request with the "security" label. We'll work with you to analyze the threats.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/job_attachments/download.py [hunk 1]
```diff
@@ -685,7 +689,14 @@ def get_job_output_paths_by_asset_root(
...
     output_manifests_by_root = get_output_manifests_by_asset_root(
-        s3_settings, farm_id, queue_id, job_id, step_id, task_id, session_action_id, session=session
+        s3_settings,
+        farm_id,
+        queue_id,
+        job_id,
+        step_id,
+        task_id,
+        session_action_id,
+        session=session,
```
Comment (godobyte) [inline src/deadline/job_attachments/download.py]: Interesting, is this updated by `hatch run fmt`? Let's make sure the formatting aligns to avoid unnecessary diffs.

File: src/deadline/job_attachments/download.py [hunk 2]
```diff
@@ -1213,3 +1233,138 @@ def download_job_output(
...
+        List[str]: Combined list of input and output manifest S3 keys
+
+    Raises:
+        JobAttachmentsError: If there's any error retrieving the manifests
+    """
+    root_prefix: str = job_attachment_settings.rootPrefix.rstrip("/")
+    output_manifest_prefix: str = f"{root_prefix}/Manifests/{farm_id}/{queue_id}/{job_id}/"
+
+    try:
+        input_manifest_keys: List[str] = _get_input_manifest_keys(
```
Comment (godobyte) [inline src/deadline/job_attachments/download.py]: nit - what about having `from_job` to following `_get_tasks_manifests_keys_from_s3` format and provide more context?

```suggestion
        input_manifest_keys: List[str] = _get_input_manifest_keys_from_job(
```

File: src/deadline/job_attachments/download.py [hunk 3]
```diff
@@ -1213,3 +1233,138 @@ def download_job_output(
...
+        )
+    except ClientError as err:
+        raise JobAttachmentsError(f"Failed to get job metadata: {str(err)}")
+
+    if "attachments" not in job_metadata or "manifests" not in job_metadata["attachments"]:
+        raise JobAttachmentsError(
+            "Job metadata missing expected attachments or manifests structure"
+        )
+
+    manifest_data: Dict[str, Any] = job_metadata["attachments"]["manifests"]
```
Comment (godobyte) [inline src/deadline/job_attachments/download.py]: Is it guaranteed the job metadata has attachments and manifests? If not, this leads to `KeyError`.

File: test/unit/deadline_job_attachments/test_download.py [hunk 4]
```diff
@@ -2546,3 +2549,354 @@ def test_get_new_copy_file_path_file_collisions(tmp_path: Path) -> None:
...
+        "DeadlineCloud/Manifests/farm-id/queue-id/Inputs/456/manifest_input",
+    ]
+    assert sorted(result) == sorted(expected)
+
+
+def test_get_input_manifest_keys_empty_manifests():
+    """Test when manifests list is empty"""
+    mock_session: MagicMock = MagicMock()
+    mock_deadline: MagicMock = MagicMock()
+    mock_deadline.get_job.return_value = {"attachments": {"manifests": []}}
```
Comment (godobyte) [inline test/unit/deadline_job_attachments/test_download.py]: `attachments` is not required, for case like this, GetJob response woudn't have `attachments` at all.

Comment (andrew-fenton) [reply test/unit/deadline_job_attachments/test_download.py]: Ah, got it. I'll remove the test case then since there's another test covering the case when no attachment key exists.

File: src/deadline/job_attachments/download.py (hunk 3 above)
Comment (andrew-fenton) [reply src/deadline/job_attachments/download.py]: Hmm, the line above checks that the keys are present but a job may have no attachments if the job has no input and output assets.

I will change the code to return a an empty list if no manifests are present.

File: src/deadline/job_attachments/download.py (hunk 1 above)
Comment (andrew-fenton) [reply src/deadline/job_attachments/download.py:699]: Ah this may have been my IDE. Will fix.

File: src/deadline/job_attachments/download.py (hunk 2 above)
Comment (andrew-fenton) [reply src/deadline/job_attachments/download.py]: Good point. I changed the name to `_get_input_manifest_keys_for_job`. I think "for" makes more sense than "from" here.

File: src/deadline/job_attachments/download.py [hunk 5]
```diff
@@ -1213,3 +1214,141 @@ def download_job_output(
         progress_tracker.total_time = time.perf_counter() - start_time
 
         return progress_tracker.get_download_summary_statistics(downloaded_files_paths_by_root)
+
+
+def _download_job_manifests_using_s3_keys(
+    session: boto3.Session,
+    manifest_keys: List[str],
+    job_attachment_settings: JobAttachmentS3Settings,
+    download_directory: str,
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py]: str or Path are both good options. Just something to consider in the code.

File: src/deadline/job_attachments/download.py [hunk 6]
```diff
@@ -1213,3 +1214,141 @@ def download_job_output(
...
+        job_attachment_settings: S3 job attachments settings
+        download_directory: Local directory path where manifests will be saved
+
+    Raises:
+        JobAttachmentsError: If manifest keys are malformed or if there are IO errors
+        JobAttachmentS3BotoCoreError: If there are AWS S3 errors (excluding missing files)
+    """
+    s3 = get_s3_client(session=session)
+
+    for manifest_key in manifest_keys:
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py]: Is this a a barebones version for downloading? We can be more optimal using Concurrent and S3 transfer manager.

File: src/deadline/job_attachments/download.py [hunk 7]
```diff
@@ -1213,3 +1214,141 @@ def download_job_output(
...
+        List[str]: Combined list of input and output manifest S3 keys
+
+    Raises:
+        JobAttachmentsError: If there's any error retrieving the manifests
+    """
+    root_prefix: str = job_attachment_settings.rootPrefix.rstrip("/")
+    output_manifest_prefix: str = f"{root_prefix}/Manifests/{farm_id}/{queue_id}/{job_id}/"
+
+    try:
+        input_manifest_keys: List[str] = _get_input_manifest_keys_for_job(
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py]: I like how you split the input and output manifests!

File: src/deadline/job_attachments/download.py [hunk 8]
```diff
@@ -1213,3 +1214,141 @@ def download_job_output(
         progress_tracker.total_time = time.perf_counter() - start_time
 
         return progress_tracker.get_download_summary_statistics(downloaded_files_paths_by_root)
+
+
+def _download_job_manifests_using_s3_keys(
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py]: Can we consider a better file home for these functions?

Download is really generic and we need to organize this file especially. Its a rats nest of code right now.

File: src/deadline/job_attachments/download.py (hunk 6 above)
Comment (andrew-fenton) [reply src/deadline/job_attachments/download.py]: Yes, this is just the simple implementation for now but will be later optimized with concurrency and transfer manager like it's done for `_download_files_parallel`.

File: src/deadline/job_attachments/download.py (hunk 8 above)
Comment (andrew-fenton) [reply src/deadline/job_attachments/download.py]: True. I created `manifest_handling.py`, `manifest_download.py`, and relocated the new functions.

File: src/deadline/job_attachments/manifest_handling.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    if "attachments" not in job_metadata or "manifests" not in job_metadata["attachments"]:
+        return []
+
+    manifest_data: List[Dict[str, Any]] = job_metadata["attachments"]["manifests"]
+
+    manifest_keys: List[str] = []
+    for manifest in manifest_data:
+        if "inputManifestPath" not in manifest:
+            # Skip if there's no input manifest path - job may have no input assets
+            continue
```
Comment (godobyte) [inline src/deadline/job_attachments/manifest_handling.py:103]: Good addition to test coverage for the edge cases!

**Review by godobyte — APPROVED** [review]

**Comment by godobyte** [comment]
Nit - Would you mind squashing these commits into a single commit per PR? (Unless having multiple commits was intentional)

**Review by leongdl — APPROVED** [review]
